### PR TITLE
refactor: reduce scope of internal keys

### LINKED ISSUE #334: make the INTERNAL_KEY_PREFIX configurable
We run into a problem. We are trying to use `raft-engine` as our Kv store. However, Our existing business logic uses the keys with the prefix `__`.  How about let's make the INTERNAL_KEY_PREFIX configurable?

### PR DESCRIPTION
**Pull request by WenyXu**
Reduce the scope of internal keys 

close #334

### CODE REVIEW

**Comment by WenyXu** [comment]
Hi @tabokie PTAL

File: src/log_batch.rs [hunk 1]
```diff
@@ -1562,12 +1562,9 @@ mod tests {
...
-                .put(0, crate::make_internal_key(&[0]), b"v".to_vec())
-                .unwrap_err(),
-            Error::InvalidArgument(_)
-        ));
+        assert!(batch
+            .put(0, crate::make_internal_key(&[0]), b"v".to_vec())
+            .is_ok());
         assert!(matches!(
             batch
                 .put_message(0, crate::make_internal_key(&[1]), &Entry::new())
```
Comment (tabokie) [inline src/log_batch.rs:1570]:
```suggestion
                .put_message(0, crate::make_internal_key(ATOMIC_GROUP_KEY), &Entry::new())
```

File: src/log_batch.rs [hunk 2]
```diff
@@ -994,7 +995,6 @@ fn verify_checksum_with_signature(buf: &[u8], signature: Option<u32>) -> Result<
 lazy_static! {
     static ref ATOMIC_GROUP_ID: Arc<AtomicU64> = Arc::new(AtomicU64::new(0));
 }
-const ATOMIC_GROUP_KEY: &[u8] = &[0x01];
```
Comment (tabokie) [inline src/log_batch.rs:997]: We should keep it here. Except make it `pub(crate)`

File: src/log_batch.rs [hunk 3]
```diff
@@ -728,7 +729,7 @@ impl LogBatch {
 
     /// Adds a protobuf key value pair into the log batch.
     pub fn put_message<S: Message>(&mut self, region_id: u64, key: Vec<u8>, s: &S) -> Result<()> {
-        if crate::is_internal_key(&key, None) {
+        if crate::is_internal_key(&key, Some(ATOMIC_GROUP_KEY)) {
```
Comment (tabokie) [inline src/log_batch.rs:732]: Don't change the callsite. Implement two versions of `is_internal_key` instead:

For `cfg(test)`, keep the behavior the same as before.

For `not(cfg(test))`, we only check `ATOMIC_GROUP_KEY` even for `is_internal_key(key, None)`. And we assert all key prefix passed to `make_internal_key` is `ATOMIC_GROUP_KEY`.

File: src/lib.rs [hunk 4]
```diff
@@ -215,5 +216,6 @@ mod tests {
         assert!(crate::is_internal_key(&key, None));
         assert!(crate::is_internal_key(&key, Some(&[0])));
         assert!(!crate::is_internal_key(&key, Some(&[1])));
+        assert!(!crate::is_internal_key(&key, Some(ATOMIC_GROUP_KEY)));
```
Comment (tabokie) [inline src/lib.rs:219]: Remove this line. This line succeeds with or without this PR.

**Review by tabokie — COMMENTED** [review]
Don't forget to update CHANGELOG.md, this should be filed under "Behavior Changes"

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,9 @@
 
 ## [Unreleased]
 
+### Behavior Changes
+* `LogBatch::put` and `LogBatch::put_message` allow to put key start with `__` but not equals the `__{ATOMIC_GROUP_KEY}`(0x95 0x95 0x01).
```
Comment (WenyXu) [inline CHANGELOG.md:6]:
```suggestion
* `LogBatch::put` and `LogBatch::put_message` allow to put key start with `__` but not equals the `__{ATOMIC_GROUP_KEY}`(0x5f 0x5f 0x01).
```

File: src/log_batch.rs [hunk 6]
```diff
@@ -728,6 +728,14 @@ impl LogBatch {
 
     /// Adds a protobuf key value pair into the log batch.
     pub fn put_message<S: Message>(&mut self, region_id: u64, key: Vec<u8>, s: &S) -> Result<()> {
+        #[cfg(not(test))]
```
Comment (tabokie) [inline src/log_batch.rs:731]: No, this is not what I meant. I mean this:
```
#[not(cfg(test))]
fn is_internal_key(key, prefix) {
  if let Some(p) = prefix { ... }
  else {
    return is_internal_key(key, Some(ATOMIC_GROUP_KEY));
  }
}
#[not(cfg(test))]
fn make_internal_key(key, prefix) {
  assert(prefix == ATOMIC_GROUP_KEY);
  ...
}
```

File: src/engine.rs [hunk 7]
```diff
@@ -1520,7 +1520,7 @@ pub(crate) mod tests {
             let mut total = 0;
             engine
                 .scan_raw_messages(rid, None, None, false, |k, _| {
-                    assert!(!crate::is_internal_key(k, None));
+                    assert!(!crate::is_internal_key(k, Some(ATOMIC_GROUP_KEY)));
```
Comment (tabokie) [inline src/engine.rs:1523]: Should not change any callsite of `is_internal_key`.

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,6 +2,9 @@
 
 ## [Unreleased]
 
+### Behavior Changes
+* `LogBatch::put` and `LogBatch::put_message` allow to put key start with `__` but not equals the `__{ATOMIC_GROUP_KEY}`(0x5f 0x5f 0x01).
```
Comment (tabokie) [inline CHANGELOG.md:6]: I think this is too verbose, we have never described the exact internal prefix before.

How about "Reduce the scope of keys reserved for internal use."

File: src/log_batch.rs (hunk 6 above)
Comment (WenyXu) [reply src/log_batch.rs:731]: Got it

**Review by tabokie — APPROVED** [review]
Thanks!

**Pull request merged**